File: src/main/java/com/anod/appwatcher/backup/JsonArrayIterator.java
package com.anod.appwatcher.backup;

import com.android.util.JsonReader;

import java.io.IOException;
import java.io.Reader;
import java.util.Iterator;

import info.anodsplace.android.log.AppLog;

/**
 * @author algavris
 * @date 29/04/2017.
 */

public class JsonArrayIterator<T> implements Iterator<T> {
    private final JsonReader mJsonReader;
    private final ObjectJsonReader<T> mObjectReader;
    private boolean mStarted;

    interface ObjectJsonReader<O> {
        O read(JsonReader reader) throws IOException;
    }

    public JsonArrayIterator(Reader reader, ObjectJsonReader<T> objectReader) {
        mJsonReader = new JsonReader(reader);
        mObjectReader = objectReader;
    }

    public void close() throws IOException {
        mJsonReader.close();
    }

    @Override
    public boolean hasNext() {
        boolean hasNext = false;
        try {
            if (!mStarted) {
                mJsonReader.beginArray();
                mStarted = true;
            }
            hasNext = mJsonReader.hasNext();

            if (!hasNext) {
                mJsonReader.endArray();
            }
        } catch (IOException e) {
            AppLog.e(e);
        }

        return hasNext;
    }

    @Override
    public T next() {
        T object = null;
        try {
            object = mObjectReader.read(mJsonReader);
        } catch (IOException e) {
            AppLog.e(e);
        }

        return object;
    }

    @Override
    public void remove() {

    }
}

File: src/main/java/com/anod/appwatcher/backup/JsonArrayIterator.kt
package com.anod.appwatcher.backup

import info.anodsplace.framework.AppLog
import info.anodsplace.framework.json.JsonReader
import java.io.IOException
import java.io.Reader

/**
 * @author Alex Gavrishev
 * *
 * @date 29/04/2017.
 */

class JsonArrayIterator<T>(reader: Reader, private val mObjectReader: JsonArrayIterator.ObjectJsonReader<T>) : Iterator<T> {
    private val mJsonReader: JsonReader = JsonReader(reader)
    private var mStarted: Boolean = false

    interface ObjectJsonReader<O> {
        @Throws(IOException::class)
        fun read(reader: JsonReader): O
    }

    @Throws(IOException::class)
    fun close() {
        mJsonReader.close()
    }

    override fun hasNext(): Boolean {
        var hasNext = false
        try {
            if (!mStarted) {
                mJsonReader.beginArray()
                mStarted = true
            }
            hasNext = mJsonReader.hasNext()

            if (!hasNext) {
                mJsonReader.endArray()
            }
        } catch (e: IOException) {
            AppLog.e(e)
        }

        return hasNext
    }

    override fun next(): T {
        return mObjectReader.read(mJsonReader)
    }
}
